### PR TITLE
add an exception case

### PR DESCRIPTION
**Pull request by chchliang**
Signed-off-by: chchliang <chen.chuanliang@zte.com.cn>

Add an exception case in this testcase.

### CODE REVIEW

**Comment by chchliang** [comment]
@GordonTheTurtle

there is no association with changes here.

04:39:00 --- FAIL: TestStateTimeoutWait (0.20s)
04:39:00 	state_test.go:137: Stop callback doesn't fire in 200 milliseconds
04:39:00 FAIL
04:39:00 coverage: 28.0% of statements
04:39:00 FAIL	github.com/docker/docker/container	0.225s

File: opts/env_test.go [hunk 1]
```diff
@@ -18,13 +18,15 @@ func TestValidateEnv(t *testing.T) {
 		"env3=abc!qwe":        "env3=abc!qwe",
 		"env_4=value 4":       "env_4=value 4",
 		"PATH":                fmt.Sprintf("PATH=%v", os.Getenv("PATH")),
+		"=a":                  fmt.Sprintf("invalid environment variable: %s", "=a"),
```
Comment (AkihiroSuda) [inline opts/env_test.go]: It is really weird to let `valids` map contain invalid values.
Let's use structs.

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

**Comment by chchliang** [comment]
func ValidateEnv(val string) (string, error) {
	arr := strings.Split(val, "=")
	if arr[0] == "" {
		return "", errors.Errorf("invalid environment variable: %s", val) //here never reach in this testcase
	}
	if len(arr) > 1 {
		return val, nil
	}
	if !doesEnvExist(val) {
		return val, nil
	}
	return fmt.Sprintf("%s=%s", val, os.Getenv(val)), nil
}

In this method `ValidateEnv`,i see this test case never reach the exception branch,so i add some  invalid values to map.

**Comment by chchliang** [comment]
Ok,i use struct replace map, thanks.
@AkihiroSuda

File: opts/env_test.go [hunk 2]
```diff
@@ -8,35 +8,110 @@ import (
...
-		"env_4=value 4":       "env_4=value 4",
-		"PATH":                fmt.Sprintf("PATH=%v", os.Getenv("PATH")),
-		"PATH=something":      "PATH=something",
-		"asd!qwe":             "asd!qwe",
-		"1asd":                "1asd",
-		"123":                 "123",
-		"some space":          "some space",
-		"  some space before": "  some space before",
-		"some space after  ":  "some space after  ",
+	valids := []struct {
```
Comment (AkihiroSuda) [inline opts/env_test.go]: testCases := []struct{
  value string
  expected string
  err error
}

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: opts/env_test.go [hunk 3]
```diff
@@ -8,35 +8,113 @@ import (
...
+			value:    "env_4=value 4",
+			expected: "env_4=value 4",
+		},
+		{
+			value:    "PATH",
+			expected: fmt.Sprintf("PATH=%v", os.Getenv("PATH")),
+		},
+		{
+			value:    "=a",
+			expected: fmt.Sprintf("invalid environment variable: %s", "=a"),
```
Comment (AkihiroSuda) [inline opts/env_test.go]: When `err` is nil, `expected` should be empty

File: opts/env_test.go [hunk 4]
```diff
@@ -8,35 +8,111 @@ import (
...
+	for _, r := range testcase {
+		actual, err := ValidateEnv(r.value)
 		if err != nil {
+
+			if err.Error() == r.err.Error() {
+				continue
+			}
+
 			t.Fatal(err)
 		}
```
Comment (AkihiroSuda) [inline opts/env_test.go]: if err != r.err { t.Fatalf("Expected err ...", ...) }

File: opts/env_test.go [hunk 5]
```diff
@@ -8,35 +8,108 @@ import (
...
-		actual, err := ValidateEnv(value)
+
+	for _, r := range testcase {
+		actual, err := ValidateEnv(r.value)
 		if err != nil {
-			t.Fatal(err)
+			if err.Error() != r.err.Error() {
+				t.Fatalf("Expected err[%v], got err[%v]", r.err, err)
+			}
 		}
```
Comment (AkihiroSuda) [inline opts/env_test.go]: Make sure it fails when `err == nil && r.err != nil`

**Review by AkihiroSuda — APPROVED** [review]
LGTM if green

**Comment by tonistiigi** [comment]
LGTM

**Review by yongtang — APPROVED** [review]
LGTM

**Pull request merged**